### DISCUSSION TITLE
Additional smart meter script for ZPA GH305 (SML)

### DISCUSSION THREAD

**Discussion by HarryKoch042018** [arendst/Tasmota#18091] [Q&A]
Dear community,
I started to enable my tasmota with a connected IR module to read the data from my energy meter ZPA GH305 (SML).
The needed script is described in a good way [here](https://tasmota.github.io/docs/Smart-Meter-Interface/#zpa-gh305-sml).
But only the values for the overall energy consumption is documented, but not the current load measured in Watt.
So, I checked the data stream that comes over IR sensor and I found out that the value
77070100100700FF@1,Actual load,W,Power_curr,0 shows the actual load.
Unfortunately I could not get a proper data sheet for this smart meter to proove my assumption, but I am pretty sure that this fits.
So, maybe the authors of the website would like to enhance the documentation with my additional row.
Many thanks
Best regards
Eckhard

**Comment by barbudor** [comment]
You can suggest changes and enhancements in the documentation with the pencil button to the top right of the page
Thanks

**Comment by marwonn** [comment]
Just want to let you know that it works. Just a slight adjustment:
1,77070100100700ff@1,Actual load,W,Power_curr,0

**Reply by HarryKoch042018** [reply]
Yes thanks - I just proposed it in the article itsself - hopefully the authors accept this new enhancement.
many thanks for your feedback

**Comment by NCO3** [comment]
> 1,77070100100700ff@1
Is this true for the **ZPA GH305.D-S1-00.00-12S** ?
Because I don't get anything but the total consumption.

**Reply by PaulHamburg** [reply]
…same here: 
ZPA GH305.D-S1-00.00-12S
only total consumption.

any idea?

**Comment by 4Jordy** [comment]
Hi Paul,
you are from Hamburg as well with the same net supplier. So I guess we will have the same Issue.
Fact is that I can only read the total amount of KWh.
So I checked the sensor.
21:56:56.503 CMD: sensor53 d1
21:56:56.508 RSL: RESULT = {"Time":"2023-08-20T21:56:56","SML":{"CMD":"dump: 1"}}
21:56:56.599 1b 1b 1b 1b 01 01 01 01 76 05 01 02 03 04 62 00 62 00 72 65 00 00 01 01 76 01 0b 0a 01 5a 50 41 00 01 88 ad 0b 05 01 02 03 
21:56:56.631 : 77 01 0b 0a 01 5a 50 41 00 01 88 ad 0b 07 01 00 62 0a ff ff 72 62 01 65 00 43 c5 20 73 
21:56:56.650 : 77 07 01 00 60 32 01 01 01 01 01 01 04 5a 50 41 01 
21:56:56.675 : 77 07 01 00 60 01 00 ff 01 01 01 01 0b 0a 01 5a 50 41 00 01 88 ad 0b 01 
21:56:57.577 : 77 07 01 00 01 08 00 ff 65 00 1c 00 04 01 62 1e 52 03 69 00 00 00 00 00 00 01 df 01 01 01 63 ab 61 00 76 05 01 02 03 04 
21:56:57.607 : 77 01 0b 0a 01 5a 50 41 00 01 88 ad 0b 07 01 00 62 0a ff ff 72 62 01 65 00 43 c5 21 73 
21:56:57.626 : 77 07 01 00 60 32 01 01 01 01 01 01 04 5a 50 41 01 
21:56:57.652 : 77 07 01 00 60 01 00 ff 01 01 01 01 0b 0a 01 5a 50 41 00 01 88 ad 0b 01 
21:56:58.602 : 77 07 01 00 01 08 00 ff 65 00 1c 00 04 01 62 1e 52 03 69 00 00 00 00 00 00 01 df 01 01 01 63 b5 00 00 76 05 01 02 03 04 
21:56:58.632 : 77 01 0b 0a 01 5a 50 41 00 01 88 ad 0b 07 01 00 62 0a ff ff 72 62 01 65 00 43 c5 22 73 

What I can se is the "77 07 01 00 01 08 00 ff" comming in for Total but what is the "77 07 01 00 60 01 00 ff 01" can somebody help what this masure ist?
Thanks
Jordy

**Reply by NCO3** [reply]
I actually can confirm that I can read the actual power in W as well as the total consumption on my ZPA GH305.D-S1-00.00-12S.
As far as I remember I had to activate the power meter with the PIN from the net supplier (in my case Hamburg as well :-)

**Reply by PaulHamburg** [reply]
Hi Jordy, did some more research for GH305. The problem here was the (horrible!) unlocking procedure with a flashlight. After entering the pin, you still have to deactivate the pin request (Pin-Abfrage: OFF) and activate the extended record (Erweiterter Datensatz: ON). Here is a usable description of this procedure : 
https://www.ww-netz.com/fileadmin/WWN/Messwesen/ZPA_GH305.pdf

With the following script, the total consumption (kWh with 3 digits after the decimal point) and the current power (W) should then be provided:

>D
>B
->sensor53 r
>M 1
+1,3,s,0,9600,SML
1,77070100010800ff@1000,Total Verbrauch,KWh,Total_in,3
1,77070100100700ff@1,Verbrauch aktuell,W,Power_curr,0
1,=h
#

Hope that fits ;-)
With best regards from Altona,
Paul

**Reply by varac** [reply]
Hej, I have the same smart meter, and want to activate the extended record - how did you get your PIN ? I didn't get any afaik.

**Reply by NCO3** [reply]
> how did you get your PIN

You can request this from your power supplier whos responsible for your meter

**Reply by 4Jordy** [reply]
> > how did you get your PIN
> 
> You can request this from your power supplier whos responsible for your meter

Yes, thant's right. In HH you can request this by email.

**Reply by Radoom** [reply]
Guy, this helped a lot, thanks.

Not sure, but maybe the PIN can be "guessed" from the meter number.  I would think that those grid operators / meter makers have built a simple approach for field technicians to figure the pin without database access when in a network-free basement room.

In my case, I can spot the PIN-digits in the meter number like this:
PIN digit 1 = 2nd-last digit from meter number
PIN digit 2 = 8th-last 
PIN digit 3 = 3rd-last
PIN digit 4 = 7th-last
So, if Meter Number is
0ZPA0087654321
Pin will be 2837
Or, if Meter Number is
0ZPA0024000310
Pin would be 1234
Maybe someone can have a check to verify or invalidate my thesis.

And another hint, maybe not clear to everyone immediately (took me two rounds to understand the procedures...)
After entering the pin, "cycle" through the display settings with short flashes until you get to "Inf" which will be "off" -> set it to "ON" with a long flash. Then Short-Flash again to get to "PIN" which will show "ON"; again, long-flash to change this to "OFF" so that the meter does not lock itself again.

And another one: Tried with a LED flashlight, that didn't work reliably, sometimes the pin digits jumped too often, sometimes not at all, or quickly to the next digit. Tried some rounds, then got the decades old 9V LEGO(tm) stuff out as I have no "normal" flashlight any more. That worked very precisely, just swiping the small lamp to the sensor and off again. Maybe the sensor likes the full light spectrum of a good old "bulp" better than LED light. Or it was the lighted area, so maybe it helps if you give your LED flashlight a "mask" (from some cardboard) leaving only a very small light dot coming out. But LEGO was more fun :-D

**Reply by varac** [reply]
> Maybe someone can have a check to verify or invalidate my thesis.

I requested a PIN for my device and your thesis does not work for me.

**Reply by Radoom** [reply]
Thanks for checking. Life could be easy sometimes... could. But maybe you can check your pin against your number and see if the digits are in. Maybe there's some rule that depends on another digit to tell us which digits are then to be used as pin, e.g. if fourth behind ZPA is 0, use my digits, if it is 1, then your's. I bet it's in there somewhere... or is the pin always the same? Mine starts with 32..

**Comment by 4Jordy** [comment]
Hey Paul,

thanks for the hint. I had to reopen the device with the pin again. So it was still closed.
I do the pin activation as well.
Thanks for the help

with kind regards from Barenfeld ;-)
Jordy

**Comment by thobastian** [comment]
Hey Paul, 
also thanks for your hint. It works in Hamburg-Hausbruch. 
:-)
Thomas

**Comment by Radoom** [comment]
See https://github.com/arendst/Tasmota/discussions/18091#discussioncomment-9916742 for some idea on the PIN number and 2 other hints that may help.

**Comment by Radoom** [comment]
Another question: in the Docs, there is the script for this meter [here](https://tasmota.github.io/docs/Smart-Meter-Interface/#zpa-gh305-sml). But I get only values for current power, and total consumed power (have no feed-in yet, so that is zero, but would likely give proper values if I would ever feed-in over this meter). But for all other values, I only get constant ZERO. Does the meter not provide these values? I have enabled the "Inf" setting (it is "ON") - would expect that at that it should deliver all the values, how is that with you guys?

**Reply by Timme2K** [reply]
Hey Radoom,
same here. I only get values for "SML Total Verbrauch", "SML Total Einspeisung" and "SML Actual Load" back. The rest of the list ist filled with 0.
I have no idea why. I also deactivated Pin and enabled "Inf".
I hope someone can help us to find a solution. Mayby the OBIS-Values are not correct, but i don't find an ducumentation for these yet.
When somone can help us, I would much appreciate it.
Best regards
Tim

**Reply by LARSMA** [reply]
Hey Tim,

I’m using the same smart meter as you and also running Tasmota. It’s been working fine for over a year – I get values like SML Total Verbrauch and SML Actual Load without issues.

However, I don’t get any value for SML Total “Einspeisung”, and that’s what I’m really trying to figure out right now.

Would you mind sharing the script you’re using in Tasmota? I’d like to try it on my setup and see if it helps to get the “Einspeisung” values working.

I also have a theory I’d like to test: maybe the smart meter is configured differently depending on whether you told your utility provider that you’re feeding solar power into the grid or not. It could be that the meter is set up not to transmit those “Einspeisung” values unless that feature is “activated” or enabled by them.

Thanks a lot in advance – would really appreciate your help!

Best regards,
Lars

**Reply by Timme2K** [reply]
Hi Lars,
I use the script for the ZPA GH305 from Tasmota.github page.

>D
>B
->sensor53 r
>M 1
+1,3,s,0,9600,SML
1,77070100010800ff@1000,Total Verbrauch,kWh,Total_in,3
1,77070100020800ff@1000,Total Einspeisung,kWh,Total_out,3
1,=h==================
1,77070100100700ff@1,Actual load,W,Power_curr,0
1,=h==================
1,=m 9+10+11 @1,Currents L1+L2+L3,A,Curr_summ,3
1,=m 12+13+14/#3 @1,Voltage L1+L2+L3/3,V,Volt_avg,3
1,=h==================
1,77070100240700ff@1,Consumption P1,W,Power_p1,2
1,77070100380700ff@1,Consumption P2,W,Power_p2,2
1,770701004c0700ff@1,Consumption P3,W,Power_p3,2
1,=h   ----
1,770701001f0700ff@1,Current L1,A,Curr_p1,3
1,77070100330700ff@1,Current L2,A,Curr_p2,3
1,77070100470700ff@1,Current L3,A,Curr_p3,3
1,=h   ----
1,77070100200700ff@1,Voltage L1,V,Volt_p1,3
1,77070100340700ff@1,Voltage L2,V,Volt_p2,3
1,77070100480700ff@1,Voltage L3,V,Volt_p3,3
1,=h==================
1,77070100510701ff@1,Phaseangle L2-L1,deg,phase_angle_L2_L1,0
1,77070100510702ff@1,Phaseangle L3-L1,deg,phase_angle_L3_L1,0
1,77070100510704ff@1,Phaseangle I/U L1,deg,phase_angle_L1,1 
1,7707010051070fff@1,Phaseangle I/U L2,deg,phase_angle_L2,1  
1,7707010051071aff@1,Phaseangle I/U L3,deg,phase_angle_L3,1 
1,770701000e0700ff@1,Frequency,Hz,Freq,1
;1,=h=======UNBEKANNT===========
;1,77070100000009FF@#,unbekannt1,,Power_Use_Sum,3
;1,77070100000000FF@#,unbekannt2,,Power_Use_Sum,3
;1,7707010060320101@#,unbekannt3,,Meter_id,0
;1,77010b0a01445a47@#,unbekannt4,,Meter_id,0
;1,77070100600100ff@#,unbekannt5,,Meter_id,0
1,=h
#

I think your theory sounds quite reasonable. My PV system is registered at my utility provider. If it is available/activated at your smart meter, it should be possible to show it directly at you smart meter display via flashlight signal in menu 2.8.0, I guess.

Hope it helps ;)

Best regards,
Tim

**Reply by LARSMA** [reply]
Hey Tim,

Thanks for your response – it really helped me dig a bit deeper and confirm my theory about the ZPA GH305 smart meter and its ability to report feed-in (exported) energy via the infrared interface, especially when trying to read it using Tasmota.

After digging into the full technical manual and doing some additional research, I’ve found that whether the meter actually provides feed-in data (like 2.8.0 for exported energy) depends entirely on the specific measurement mode the meter was configured for.

There are four main variants defined in the manual:
	1.	+A with reverse lock – this mode only counts consumption. If you generate excess power (e.g. from a balcony solar panel), the meter just ignores it. No reverse flow is recorded or transmitted.
	2.	+A | -A – this is the bidirectional mode. Both import and export are measured and provided (e.g. registers 1.8.0 for consumption and 2.8.0 for feed-in).
	3.	-A with reverse lock – export-only mode, rarely used in households.
	4.	-A netting mode (without reverse lock) – net power flow is recorded.

The catch is: you can’t change this mode yourself. It’s determined either by the firmware version or by the utility/Metering Point Operator (MSB) during installation. The mode is indicated in the meter’s type designation (something like GH305.D-...), where the key digit (usually right after the D) tells you which variant you have:
1 = +A only (no feed-in data),
2 = +A | -A (bidirectional),
3 or 4 = export-focused or netting.

So even though the hardware technically supports feed-in measurement, if your specific unit is configured as variant 1, then it’s factory-set to only register consumption – and won’t output any feed-in values via the infrared interface, no matter what Tasmota or its script tries to read. The limitation isn’t with Tasmota itself, but with the meter’s locked-in mode determined at the time of installation.

At least now I know that the limitation lies within the meter itself and not my Tasmota script – so I’ll just have to find a workaround that fits my specific use case.

Best regards,
Lars

**Comment by lassediercks** [comment]
Moinsen,

ich bin über dieses Video nochmal durch meine Settings gegangen und hab gemerkt, dass Info bei bei `off` war.

Es ist echt knifflig das auf an zu kriegen aber seitdem kann ich auch current load und nicht nur total usage auslesen

**Reply by varac** [reply]
Which Video ?

**Reply by lassediercks** [reply]
sorry I thought I posted it 
https://www.youtube.com/watch?v=HY6KSwvsZjY

**Comment by pb1966** [comment]
I have another question about reading out “Actual load”. It is not transmitted whether the current is flowing into the grid or whether it is being consumed. The meter has a +A and -A display but I can't find a way of reading this out of the data ? 

Does anyone have an idea?

**Reply by FS1961** [reply]
Does anyone have an idea of how to integrate this meter via Zigbee, Tasmota, WLAN? is there a device (not a kit) that I can purchase?

**Reply by Timme2K** [reply]
Hi PB, you are sure that you have the ZPA GH305? Maybe you have a different one and this script doesn't fit your smart meter.
Unlocking your smart meter with the pin, you already did?

Hi FS, I just use a simple IR-write-and-read-head 30€ from eBay. It runs tasmota and connects via WiFi. You can implement it to your HA via MQTT for example.
Hope it helps you.
Have a good one.
